plugins/shuffle/round.py

```diff
@@ -227,41 +239,45 @@ def process_shuffling(self):
-                    self.phase = 'BroadcastOutput'
-                    self.send_message()
-                    self.log_message("encrypt new address")
-                except CryptoError:
-                    accused = sender
-        else:
-            sender = self.players[self.previous_player()]
-            if self.inbox[phase].get(sender):
-                self.messages.packets.ParseFromString(self.inbox[phase][sender])
-                try:
-                    for packet in self.messages.packets.packet:
-                        packet.packet.message.str = self.crypto.decrypt(packet.packet.message.str)
-                    if not self.different_ciphertexts():
-                        self.messages.add_str(self.encrypt_new_address())
+            msg = self.inbox[phase].get(sender)
+            if msg:
+                accused = self._parse_from_string_safe(msg, sender)
+                if not accused:
+                    try:
+                        for packet in self.messages.packets.packet:
+                            packet.packet.message.str = self.crypto.decrypt(packet.packet.message.str)
+                        self.messages.add_str(self.addr_new)
                         self.messages.shuffle_packets()
-                        self.send_message(destination=self.players[self.next_player()])
-                        self.log_message("encrypt new address")
                         self.phase = 'BroadcastOutput'
-                    else:
+                        self.send_message()
+                        self.log_message("encrypt new address")
+                    except CryptoError:
+                        accused = sender
+        else:
+            sender = self.players[self.previous_player()]
+            msg = self.inbox[phase].get(sender)
+            if msg:
+                accused = self._parse_from_string_safe(msg, sender)
+                if not accused:
+                    try:
+                        for packet in self.messages.packets.packet:
+                            packet.packet.message.str = self.crypto.decrypt(packet.packet.message.str)
+                        if not self.different_ciphertexts():
+                            self.messages.add_str(self.encrypt_new_address())
+                            self.messages.shuffle_packets()
+                            self.send_message(destination=self.players[self.next_player()])
+                            self.log_message("encrypt new address")
+                            self.phase = 'BroadcastOutput'
+                        else:
+                            accused = sender
+                    except CryptoError:
                         accused = sender
-                except CryptoError:
-                    accused = sender
         if accused is not None:
             self.skipped_equivocation_check(accused)
-            self.log_message("wrong from " + str(accused))
+            self.log_message("bad shuffle message from " + str(accused))
```

Comment:
accused can only be `sender` with the above logic. Is that right?

Reply:
Yeah either that or None. :)